browser/ntp_sponsored_images/android/ntp_sponsored_images_bridge.cc

```diff
@@ -0,0 +1,150 @@
+}
+
+}  // namespace
+
+NTPSponsoredImagesBridge::NTPSponsoredImagesBridge(Profile* profile)
+    : profile_(profile),
+      view_counter_service_(ViewCounterServiceFactory::GetForProfile(profile_)),
+      sponsored_images_service_(
+          g_brave_browser_process->ntp_sponsored_images_service()) {
+  DCHECK_CURRENTLY_ON(BrowserThread::UI);
+
+  java_object_.Reset(Java_NTPSponsoredImagesBridge_create(
+      AttachCurrentThread(), reinterpret_cast<intptr_t>(this)));
+
+  sponsored_images_service_->AddObserver(this);
+}
+
+NTPSponsoredImagesBridge::~NTPSponsoredImagesBridge() {
+  sponsored_images_service_->RemoveObserver(this);
+  Java_NTPSponsoredImagesBridge_destroy(AttachCurrentThread(), java_object_);
+}
+
+static base::android::ScopedJavaLocalRef<jobject>
+JNI_NTPSponsoredImagesBridge_GetInstance(JNIEnv* env,
+                                      const JavaParamRef<jobject>& j_profile) {
+  auto* profile = ProfileAndroid::FromProfileAndroid(j_profile);
+  return ntp_sponsored_images::NTPSponsoredImagesBridgeFactory::GetInstance()
+             ->GetForProfile(profile)->GetJavaObject();
+}
+
+base::android::ScopedJavaLocalRef<jobject>
+NTPSponsoredImagesBridge::GetJavaObject() {
+  return base::android::ScopedJavaLocalRef<jobject>(java_object_);
+}
+
+void NTPSponsoredImagesBridge::RegisterPageView(
+    JNIEnv* env, const JavaParamRef<jobject>& obj) {
+  DCHECK_CURRENTLY_ON(BrowserThread::UI);
+  if (view_counter_service_)
+    view_counter_service_->RegisterPageView();
+}
+
+base::android::ScopedJavaLocalRef<jobject>
+NTPSponsoredImagesBridge::CreateWallpaper() {
+  JNIEnv* env = AttachCurrentThread();
+
+  auto data = view_counter_service_->GetCurrentWallpaperForDisplay();
+  if (!view_counter_service_ || data.is_none())
+    return base::android::ScopedJavaLocalRef<jobject>();
+
+  // TODO(bridiver) - need to either expose these constants or change this
+  // to a struct instead of base::Value
+  auto* image_path = data.FindStringPath("wallpaperImagePath");
+  auto* logo_image_path = data.FindStringPath("logo.imagePath");
+  if (!image_path || !logo_image_path)
+    return base::android::ScopedJavaLocalRef<jobject>();
+
+  auto focal_point_x = data.FindIntPath("focalPoint.x");
+  auto focal_point_y = data.FindIntPath("focalPoint.y");
+  auto* logo_destination_url = data.FindStringPath("logo.destinationUrl");
+
+  return Java_NTPSponsoredImagesBridge_createWallpaper(
+      env,
+      ConvertUTF8ToJavaString(env, *image_path),
+      focal_point_x ? *focal_point_x : 0,
+      focal_point_y ? *focal_point_y : 0,
+      ConvertUTF8ToJavaString(env, *logo_image_path),
+      ConvertUTF8ToJavaString(env, logo_destination_url ? *logo_destination_url
+                                                        : ""));
+}
+
+base::android::ScopedJavaLocalRef<jobject>
+NTPSponsoredImagesBridge::GetCurrentWallpaper(
+    JNIEnv* env, const JavaParamRef<jobject>& obj) {
+  DCHECK_CURRENTLY_ON(BrowserThread::UI);
+  return CreateWallpaper();
```

Comment:
How about unifying this method with above `GetCurrentWallpaper()`?

Reply:
it used to be called in two different places, but we can do that in a follow up with some other cleanup on the java side